[PORT] DB Achievements

Ivniinvi opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request
Ports DB achievements from TG.
Ports the following:
tgstation/tgstation#47058
tgstation/tgstation#47617
tgstation/tgstation#48007
tgstation/tgstation#48146
And the relevant parts of:
tgstation/tgstation#51255

**This will require migration using the enclosed tool**

<!-- Describe The Pull Request. Please be sure every change is documented or this can delay review and even discourage maintainers from merging your PR! -->

## Why It's Good For The Game
We no longer rely on the BYOND hub for achievements, instead using the DB.
<!-- Please add a short description of why you think these changes would benefit the game. If you can't justify it in words, it might not be worth adding. -->

## Changelog
:cl: Ivniinvi, AMonkeyThatCodes, Qustinnus, Willox, AnturK, SpaceManiac
server: Achievements are now handled by the DB instead of the BYOND hub
add: The verb check-achievements in the OOC tab can be used to see your achievements
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

Ivniinvi committed:
initial commit

Ivniinvi committed:
DB and Finishing Touches

Ivniinvi committed:
Merge branch 'master' of https://github.com/BeeStation/BeeStation-Hornet into cheevoport

ike709 commented:
based

qwertyquerty commented:
Bring in all the existing medals into this system

Ivniinvi commented:
> Bring in all the existing medals into this system

It has everything except G-G-Ghosts

Ivniinvi committed:
Re-adds the Ghosts medal and rewrites descriptions

Ivniinvi commented:
I've now re-added the Ghosts medal (thanks Archanial) and edited the descriptions Crossed didn't like.

qwertyquerty commented:
Alright, so I think we need a way to award the users who currently have medals their respective achievements.

Ivniinvi commented:
> Alright, so I think we need a way to award the users who currently have medals their respective achievements.

Yes, that's what HubMigrator does.

KubeRoot commented:
> I've now re-added the Ghosts medal (thanks Archanial) and edited the descriptions Crossed didn't like.

Maybe edit the PR description to remove the mention of not having that one

Ivniinvi commented:
> > I've now re-added the Ghosts medal (thanks Archanial) and edited the descriptions Crossed didn't like.
> 
> Maybe edit the PR description to remove the mention of not having that one

Oop, thanks, updated.

qwertyquerty reviewed (dismissed):
Remove your config changes

Ivniinvi committed:
oops

Ivniinvi committed:
Merge branch 'master' of https://github.com/BeeStation/BeeStation-Hornet into cheevoport

Ivniinvi commented:
![image](https://user-images.githubusercontent.com/22421329/138346857-8070e077-8890-49d2-9d48-78430d06788f.png)

Crossed has approved the portions that he is code owner of (but he didn't mark it as approved on github)

Crossedfall commented on tools/HubMigrator/HubMigrator.dm:
```diff
@@ -0,0 +1,183 @@
+//Misc Medal hub IDs
+#define MEDAL_METEOR 			"Your Life Before Your Eyes"
```
I only did a brief check, but I found 10 that didn't correctly match up. Should go back through this list and make sure it's all correct.

Crossedfall commented on code/datums/achievements/boss_achievements.dm:
```diff
@@ -0,0 +1,82 @@
+/datum/award/achievement/boss/boss_killer
+	name = "Boss Killer"
+	desc = "You've come a long ways from asking how to switch hands."
+	database_id = "Boss Killer"
```
```suggestion
	database_id = BOSS_MEDAL_ANY
```

Crossedfall commented:
> ![image](https://user-images.githubusercontent.com/22421329/138346857-8070e077-8890-49d2-9d48-78430d06788f.png)
> 
> Crossed has approved the portions that he is code owner of (but he didn't mark it as approved on github)

I haven't approved anything yet, I just told you that I would only approve the stuff I had ownership of.

Ivniinvi committed:
Addresses comments

Ivniinvi commented:
> > Crossed has approved the portions that he is code owner of (but he didn't mark it as approved on github)
> 
> I haven't approved anything yet, I just told you that I would only approve the stuff I had ownership of.

Oops, sorry, misinterpreted that

Ivniinvi commented:
Right, I've addressed the concerns presented.

Various inconsistencies:
The "Exterminator" medal exists in the code and on the hub, but no one has it.
The "Crusher" medal exists in the code but not on the hub.
The "Forgotten Star" medal exists in the code but not on the hub.
The "Avada Kedavra!" medal exists on the hub but not in the code, and no one has it.

As such, none of these four have been included in HubMigrator.dm

Crossedfall commented on tools/HubMigrator/HubMigrator.dm:
```diff
@@ -0,0 +1,166 @@
+		for(var/score in scores_data[keyv])
+			values += "('[keyv]','[score]',[scores_data[keyv][score]],now())"
+		if(values.len)
+			var/list/keyline = list("INSERT INTO [ach](ckey,achievement_key,value,last_updated) VALUES")
```
last_updated should be when the achievement was unlocked (which is already part of the scraped data).  Currently, this just sets all of the dates to the current date.

Ivniinvi committed:
pleasy worky

Ivniinvi committed:
metadata plus plus

Crossedfall commented on SQL/database_changelog.txt:
```diff
@@ -1,15 +1,43 @@
+
+Added achievement_name and achievement_description columns to achievement_metadata table.
+
+ALTER TABLE `achievement_metadata` ADD COLUMN (`achievement_name` VARCHAR(64) NULL DEFAULT NULL, `achievement_description` VARCHAR(512) NULL DEFAULT NULL);
```
```suggestion
DROP TABLE IF EXISTS `achievement_metadata`;
CREATE TABLE `achievement_metadata` (
	`achievement_key` VARCHAR(32) NOT NULL,
	`achievement_version` SMALLINT UNSIGNED NOT NULL DEFAULT 0,
	`achievement_type` enum('achievement','score','award') NULL DEFAULT NULL,
	`achievement_name` VARCHAR(64) NULL DEFAULT NULL,
	`achievement_description` VARCHAR(512) NULL DEFAULT NULL,
	PRIMARY KEY (`achievement_key`)
) ENGINE=InnoDB;
```

There's no reason to have two queries when the table doesn't exist yet.

Ivniinvi committed:
Update SQL/database_changelog.txt

Co-authored-by: Crossedfall <ping@crossedfall.io>

Crossedfall commented on SQL/database_changelog.txt:
```diff
@@ -1,15 +1,41 @@
+Added achievements table.
+See hub migration verb in _achievement_data.dm for details on migrating.
+
+CREATE TABLE `achievements` (
```
```suggestion
DROP TABLE IF EXISTS `achievements`;
CREATE TABLE `achievements` (
```

Crossedfall replied:
Consistency with the other query

Ivniinvi committed:
Update SQL/database_changelog.txt

Co-authored-by: Crossedfall <ping@crossedfall.io>

Crossedfall reviewed (approved):
Everything works now.

Crossedfall commented:
DB updated and cheevos ported.
